src/hotspot/share/runtime/objectMonitor.cpp

```diff
@@ -312,7 +312,60 @@ void ObjectMonitor::ClearSuccOnSuspend::operator()(JavaThread* current) {
 // -----------------------------------------------------------------------------
 // Enter support
 
+bool ObjectMonitor::enter_for(JavaThread* locking_thread) {
+  // Used by ObjectSynchronizer::enter_for to enter for another thread
+  // this code may only contend with deflation.
+  assert(locking_thread == Thread::current() || locking_thread->is_obj_deopt_suspend(), "must be");
+
+  // Block out deflation as soon as possible.
+  add_to_contentions(1);
+
+  bool success = false;
+  if (!is_being_async_deflated()) {
+    void* prev_owner = try_set_owner_from(nullptr, locking_thread);
+
+    if (prev_owner == nullptr) {
+      assert(_recursions == 0, "invariant");
+      success = true;
+    } else if (prev_owner == locking_thread) {
```

Comment:
I thought you would have to handle `LockingMode == LM_LEGACY && locking_thread->is_lock_owned((address)cur) == true`. Could this cause the timeouts?

Reply:
It is the ` -XX:LockingMode=0 -Xlog:monitorinflation=trace ` that timeout. 
You are correct that this case should also be handled. 
No tests reach this point. But it should still be reachable. The re-lock code will never cause inflation, but it might relock a lock which was inflated by the current thread (due to hash code) and then relock calls `enter_for`

Reply:
I'm confused. The `LockingMode == LM_LEGACY && locking_thread->is_lock_owned((address)prev_owner) == true`
case is handled below on L350 -> L355. What am I missing?

Reply:
With LM_LEGACY `ObjectMonitor::_owner` will either be the JavaThread* of the owner or the BasicLock* in the stack of the owner found in the mark word of the object when the lock gets inflated:
```c++
  void* volatile _owner;            // pointer to owning thread OR BasicLock
```
Here at L332 the JavaThread* case is handled. At L350 -> 355 the BasicLock* case is handled.

Reply:
Ummm... Here's the comment line I'm talking about:
> I thought you would have to handle `LockingMode == LM_LEGACY && locking_thread->is_lock_owned((address)prev_owner) == true`. 

and the code that I referred to on L350 is exactly that line of code. Again, what am I missing?

Reply:
Axel added that line after my comment.

Reply:
Thanks for figuring out that's what I was missing! You made the comment, Axel changed
the code well below your comment in the PR, GitHub didn't connect the comment and
the code change together to flag anything as "outdated". And most importantly, I missed
that there was a commit made in response to your comment from the history.

Reply:
I do find it difficult to follow what's going on once there are more than just a few comments and commits.